lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/client/AzureBatchJobSubmissionHandler.java

```diff
@@ -39,59 +52,129 @@
+    this.jobJarMaker = jobJarMaker;
+    this.reefFileNames = reefFileNames;
+
+    this.azureBatchAccountUri = azureBatchAccountUri;
+    this.azureBatchAccountName = azureBatchAccountName;
+    this.azureBatchAccountKey = azureBatchAccountKey;
+    this.azureBatchPoolId = azureBatchPoolId;
+
+    this.applicationId = "HelloWorldJob-"
+        + this.azureBatchAccountName + "-"
+        + (new Date()).toString()
+        .replace(' ', '-')
+        .replace(':', '-')
+        .replace('.', '-');
+  }
+
+  @Override
+  public String getApplicationId() {
+    return this.applicationId;
+  }
+
+  @Override
+  public void close() throws Exception {
+    LOG.log(Level.INFO, "Closing " + AzureBatchJobSubmissionHandler.class.getName());
+  }
+
+  @Override
+  public void onNext(final JobSubmissionEvent jobSubmissionEvent) {
+
+    LOG.log(Level.FINEST, "Submitting job: {0}", jobSubmissionEvent);
+
+    try (final AzureBatchJobSubmissionHelper helper = new AzureBatchJobSubmissionHelper(
+        this.azureBatchAccountUri,
+        this.azureBatchAccountName,
+        this.azureBatchAccountKey,
+        this.azureBatchPoolId,
+        getApplicationId())) {
 
-    @Override
-    public void onNext(final JobSubmissionEvent jobSubmissionEvent) {
-        final String id = jobSubmissionEvent.getIdentifier();
-        LOG.log(Level.FINEST, "Submitting job: {0}", jobSubmissionEvent);
-
-        try (final AzureBatchJobSubmissionHelper helper = new AzureBatchJobSubmissionHelper(
-                this.azureBatchAccountUri,
-                this.azureBatchAccountName,
-                this.azureBatchAccountKey,
-                this.azureBatchPoolId,
-                getApplicationId())) {
-            helper.submit();
-        }
-        catch (IOException e) {
-            LOG.log(Level.WARNING, "Failed to create the Driver application.");
-        }
+      final String id = jobSubmissionEvent.getIdentifier();
+
+      LOG.log(Level.FINE, "Creating a job folder on Azure.");
+      final URI jobFolderURL = this.azureUploader.createJobFolder(id);
+
+      LOG.log(Level.FINE, "Assembling Configuration for the Driver.");
+      final Configuration driverConfiguration = makeDriverConfiguration(jobSubmissionEvent, id, jobFolderURL);
+
+      LOG.log(Level.FINE, "Making Job JAR.");
+      final File jobSubmissionJarFile =
+          this.jobJarMaker.createJobSubmissionJAR(jobSubmissionEvent, driverConfiguration);
+
+      LOG.log(Level.FINE, "Uploading Job JAR to Azure.");
+      final LocalResource uploadedFile = this.azureUploader.uploadFile(jobSubmissionJarFile);
+
+      LOG.log(Level.FINE, "Assembling application submission.");
+      final String command = getCommandString(jobSubmissionEvent);
+
+      helper.submit(uploadedFile, command);
+
+    } catch (final IOException ex) {
+      LOG.log(Level.SEVERE, "Error submitting Azure Batch request", ex);
+      throw new RuntimeException(ex);
     }
```

Comment:
1. we may want to separate the error between azureUploader(blob error) and Batch error.
2. AzureBatchJobSubmissionHelper.submit() throws BatchErrorException as well. Only catch IOException may not be good enough.

Reply:
1. If we separated the errors, where and how could we take advantage of that? I thought this class is used by REEF client code which is shared across runtimes and therefore will not and cannot implement runtime-specific logic.

2. BatchErrorException is a subclass of RuntimeException, which means, it is an unchecked exception. Here we catch IOException (which is a checked exception) and re-throw that inside of an unchecked RuntimeException (the method signature in the interface prohibits checked exceptions). Catching BatchErrorException and re-wrapping that in RuntimeException would not make sense since BatchErrorException inherits from RuntimeException. Am I missing something?

Reply:
You are right. My point is more about providing accurate logging for convenience here. For 1st, I think it is ok. For 2nd, I didn't notice BatchErrorException is unchecked exception. In that case, the only difference is missing a log "LOG.log(Level.SEVERE, "Error submitting Azure Batch request", ex); " But it should be fine.

Reply:
Ok, I can add a log statement.